client/web/src/components/popovers/ColorPicker.vue

```diff
@@ -115,13 +117,197 @@
 
 <script lang="ts">
 import { defineComponent } from "vue";
+import { clamp, HSV2RGB, RGB2HSV, RGB, HSV } from "../../lib/utils";
+
+interface Rect {
+	width: number;
+	height: number;
+	left: number;
+	top: number;
+}
+
+interface ColorPickerData {
+	color: HSV;
+	hue: {
+		rect: Rect;
+	};
+	saturation: {
+		rect: Rect;
+	};
+	opacity: {
+		rect: Rect;
+	};
+}
+
+type ColorPickerState = "idle" | "move_hue" | "move_opacity" | "move_saturation";
 
 export default defineComponent({
+	colorPicker: {
+		color: {
+			h: 0,
+			s: 0,
+			v: 0,
+			a: 1,
+		},
+		hue: {
+			rect: {
+				width: 0,
+				height: 0,
+				top: 0,
+				left: 0,
+			},
+		},
+		opacity: {
+			rect: {
+				width: 0,
+				height: 0,
+				top: 0,
+				left: 0,
+			},
+		},
+		saturation: {
+			rect: {
+				width: 0,
+				height: 0,
+				top: 0,
+				left: 0,
+			},
+		},
+	} as ColorPickerData,
```

Comment:
It would be great if we can avoid unsafe casting (since `as` tells TypeScript to ignore type checks and blindly believe the code that the type cast is correct). When we need to cast something to an HTML element, that's fine of course since there is no other way around it. This applies throughout the PR in a bunch of places.